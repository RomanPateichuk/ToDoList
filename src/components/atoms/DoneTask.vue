<template>
  <div :class="$style.wrapper">
    <input
      :id="id"
      type="checkbox"
      :class="$style.done"
      v-on:click="completeTask(id)"
      :checked="getCheckedValue(id)"
    />
    <label :for="id" :class="$style.chbEmptyLabel"></label>
  </div>
</template>

<script>
import { mapGetters } from "vuex";
export default {
  computed: {
    ...mapGetters(["getCheckedValue"]),
  },
  props: {
    id: {
      type: String,
    },
  },
  methods: {
    completeTask: function (id) {
      this.$store.commit("completeTask", id);
    },
  },
};
</script>

<style lang="scss" module>
@import "@/assets/scss/main.scss";
.wrapper {
  .chbEmptyLabel {
    &:before {
      content: "";
      display: block;
      width: 1.5rem;
      height: 1.5rem;
      background: $task-done;
      border-radius: 0.3125rem;
      position: absolute;
      top: 0.8125rem;
      left: 1.25rem;
    }

    &:after {
      content: "";
      display: block;
      width: 0.589375rem;
      height: 0.856875rem;
      border: 1px solid $main;
      border-left: none;
      border-top: none;
      position: absolute;
      transform: rotate(45deg);
      top: 0.9375rem;
      left: 1.75rem;
      opacity: 0;
    }
  }
  .done {
    appearance: none;
    &:checked {
      + .chbEmptyLabel {
        &:after {
          opacity: 1;
        }
      }
    }
  }
}
</style>
