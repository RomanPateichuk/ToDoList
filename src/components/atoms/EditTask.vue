<template>
  <div :class="$style.task">
    <input
      :class="[
        $style.text,
        { [$style.active]: onActiveInput },
        { [$style.unactive]: getCheckedValue(id) },
      ]"
      ref="taskInput"
      :readonly="readonly"
      @keyup.enter="editTask"
      v-model="newTaskValue"
      :placeholder="TaskValue"
    />
    <button :class="$style.edit" v-on:click="editTask()">
      {{ editStatus }}
    </button>
  </div>
</template>

<script>
import { mapGetters } from "vuex";
export default {
  computed: {
    ...mapGetters(["getCheckedValue"]),
  },
  data() {
    return {
      readonly: true,
      editStatus: "edit",
      onActiveInput: false,
      newTaskValue: "",
    };
  },
  props: {
    TaskValue: {
      type: String,
      default: "",
    },
    id: {
      type: String,
    },
  },
  methods: {
    editTask: function () {
      this.OnActiveInput = !this.OnActiveInput;
      this.$refs.taskInput.focus();
      if (this.editStatus === "edit") {
        this.editStatus = "save";
      } else {
        this.editStatus = "edit";
        this.$store.commit("saveEditTask", {
          id: this.id,
          value: this.newTaskValue,
        });
        this.newTaskValue = "";
      }

      this.readonly = !this.readonly;
    },
  },
};
</script>

<style lang="scss" module>
@import "@/assets/scss/main.scss";
.task {
  color: $task-text;
  padding-left: 2rem;
  width: 90%;
  .text {
    background-color: inherit;
    border: none;
    width: 80%;
    padding: 0.4rem;
    &.active {
      background-color: #ffff;
    }
    &.unactive {
      text-decoration: line-through;
    }
    &::placeholder {
      color: rgb(43, 57, 61);
      font-weight: 600;
    }
  }
  .edit {
    background: $task-done;
    border-radius: 0.3125rem;
    padding: 0.5rem;
    margin-left: 0.5rem;
  }
}
</style>