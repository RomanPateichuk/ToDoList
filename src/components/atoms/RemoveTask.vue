<template>
  <button @click="removeTask(id)" :class="$style.remove">&times;</button>
</template>
<script>
export default {
  methods: {
    removeTask: function (id) {
      this.$emit("callRemoveTask", id);
    },
  },
  props: {
    id: {
      type: String,
    },
  },
};
</script>

<style lang="scss" module>
@import "@/assets/scss/main.scss";

.remove {
  display: flex;
  justify-content: center;
  width: 1.5rem;
  height: 1.5rem;
  border: 1.5px solid $task-remove;
  color: $task-remove;
  border-radius: 0.3125rem;

  &:hover {
    border: 1.5px solid $task-remove-hover;
    color: $task-remove-hover;
    cursor: pointer;
  }
}
</style>
