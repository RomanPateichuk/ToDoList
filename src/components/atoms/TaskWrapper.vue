<template>
  <div class="task">
    <DoneTaskBtn :TaskValue="TaskValue" />
    <RemoveTaskBtn />
  </div>
</template>

<script>
export default {
  components: {
    RemoveTaskBtn: () => import("@/components/molecules/RemoveTaskBtn.vue"),
    DoneTaskBtn: () => import("@/components/molecules/DoneTaskBtn.vue"),
  },
  props: {
    TaskValue: {
      type: String,
      default: "",
    },
  },
};
</script>

<style scoped lang="scss">
@import "@/assets/scss/main.scss";

.task {
  &:not(:first-child) {
    margin: 1.5625rem 0 0 0;
  }

  :last-child {
    margin-left: auto;
  }
}
</style>
