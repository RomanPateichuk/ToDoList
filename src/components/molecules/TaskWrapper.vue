<template>
  <div :class="$style.task">
    <DoneTask :id="id" @completeTask="callMuttationComplete" />
    <EditTask
      :TaskValue="TaskValue"
      :id="id"
      @callEditTask="callMuttationEdit"
    />
    <RemoveTask :id="id" @callRemoveTask="callMuttationRemove" />
  </div>
</template>

<script>
export default {
  components: {
    RemoveTask: () => import("@/components/atoms/RemoveTask.vue"),
    EditTask: () => import("@/components/atoms/EditTask.vue"),
    DoneTask: () => import("@/components/atoms/DoneTask.vue"),
  },
  props: {
    TaskValue: {
      type: String,
      default: "",
    },
    id: {
      type: String,
    },
  },

  methods: {
    callMuttationComplete: function (id) {
      this.$store.commit("completeTask", id);
    },
    callMuttationEdit: function (value, id) {
      this.$store.commit("saveEditTask", {
        id: id,
        value: value,
      });
    },
    callMuttationRemove: function (id) {
      this.$store.commit("deleteTask", id);
    },
  },
};
</script>

<style lang="scss" module>
@import "@/assets/scss/main.scss";

.task {
  @include Task;
  &:not(:first-child) {
    margin: 1.5625rem 0 0 0;
  }
}
</style>
