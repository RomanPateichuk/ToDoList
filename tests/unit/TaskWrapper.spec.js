import { shallowMount, createLocalVue } from "@vue/test-utils";
import TaskWrapper from "../../src/components/molecules/TaskWrapper.vue";
import DoneTask from "../../src/components/atoms/DoneTask.vue";
import RemoveTask from "../../src/components/atoms/RemoveTask.vue";
import EditTask from "../../src/components/atoms/EditTask.vue";
import Vuex from "vuex";

const localVue = createLocalVue();
localVue.use(Vuex);

const mutations = {
  completeTask: jest.fn(),
  deleteTask: jest.fn(),
  saveEditTask: jest.fn(),
};

const getters = {
  getCheckedValue: () => jest.fn(),
};

const store = new Vuex.Store({ mutations, getters });

describe("Тестирование компонента TaskWrapper:", () => {
  const wrapper = shallowMount(TaskWrapper, {
    store,
    localVue,
    propsData: {
      TaskValue: "Task 1",
      id: "1",
    },
  });

  test("соответствие снимку", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("вызов события completeTask из дочернего компонента", () => {
    expect(
      wrapper.findComponent(DoneTask).vm.$emit("completeTask", "1")
    ).toBeTruthy();
  });

  test("Вызов метода 'callRemoveTask' из дочернего компонента", () => {
    expect(
      wrapper.findComponent(RemoveTask).vm.$emit("callRemoveTask", "1")
    ).toBeTruthy();
  });

  test("Вызов мутации 'completeTask' c нагрузкой", () => {
    expect(mutations.completeTask).toHaveBeenCalledWith({}, "1");
  });

  test("Вызов мутации 'deleteTask' с нагрузкой", () => {
    expect(mutations.deleteTask).toHaveBeenCalledWith({}, "1");
  });

  test("Вызов мутации 'editTask' с нагрузкой", () => {
    wrapper.findComponent(EditTask).vm.$emit("callEditTask", "value", "1");
    expect(mutations.saveEditTask).toHaveBeenCalledWith(
      {},
      { id: "1", value: "value" }
    );
  });

  test("Тестирование входных параметров", () => {
    expect(
      wrapper.findComponent(EditTask).findComponent("input").attributes().placeholder
    ).toBe("Task 1");
  });
});
