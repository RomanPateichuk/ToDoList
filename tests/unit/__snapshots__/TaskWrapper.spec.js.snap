// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Тестирование компонента TaskWrapper: соответствие снимку 1`] = `
<div>
  <div><input id="1" type="checkbox"> <label for="1"></label></div>
  <div><input readonly="readonly" placeholder="Task 1" class=""> <button>
      edit
    </button></div> <button>×</button>
</div>
`;
